### PR TITLE
rgw/simplefile: file based backend for rgw

### PR DESCRIPTION
**Pull request by jecluis**
This PR relies on @irq0's work.

~**NOTE:** there are a few commits in this patch set that need to be either reworked or dropped before merging. One in particular actually changes RGW behavior to make the store work.~

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

### CODE REVIEW

**Comment by votdev** [comment]
Are there plans to add a compiler flag to enabled/disable the SimpleFileStore like `WITH_RADOSGW_DBSTORE`?

**Comment by jecluis** [comment]
I believe it's already there. `WITH_RADOSGW_SIMPLEFILE` I think.

File: src/rgw/rgw_sal.cc [hunk 1]
```diff
@@ -118,6 +121,12 @@ rgw::sal::Store* StoreManager::init_storage_provider(const DoutPrefixProvider* d
     }
     return store;
   }
+  else if (svc.compare("simplefile") == 0) {
```
Comment (votdev) [inline src/rgw/rgw_sal.cc:124]: Does it make sense to encasulate this in a `WITH_RADOSGW_SIMPLEFILE` `ifdef`?

File: src/rgw/rgw_sal.cc [hunk 2]
```diff
@@ -211,6 +220,10 @@ rgw::sal::Store* StoreManager::init_raw_storage_provider(const DoutPrefixProvide
     }
   }
 
+  if (svc.compare("simplefile") == 0) {
```
Comment (votdev) [inline src/rgw/rgw_sal.cc:223]: Does it make sense to encasulate this in a `WITH_RADOSGW_SIMPLEFILE` `ifdef`?

File: src/rgw/rgw_main.cc [hunk 3]
```diff
@@ -374,6 +395,10 @@ int radosgw_Main(int argc, const char **argv)
   }
 #endif
 
+  if (config_store == "simplefile") {
```
Comment (votdev) [inline src/rgw/rgw_main.cc:398]: Does it make sense to encasulate this in a WITH_RADOSGW_SIMPLEFILE ifdef?

**Review by votdev — CHANGES_REQUESTED** [review]

**Comment by giubacc** [comment]
have we got an image on github to be pulled for doing some tests?

**Comment by jecluis** [comment]
not really, not yet. I pushed this out and went away because my brain felt like mush. I'll take care of that this morning.

**Comment by 0xavi0** [comment]
How did you test it?
do you have a script or something so we see what can be done and what is still not working?
I'm asking because I've built your branch and got a couple of cores. 
I saw you're using a dummy user with keys being 'test', but I get 403 errors when creating a bucket, for example.

**Comment by votdev** [comment]
Thanks for adding comments. I'm missing that in the whole RGW code. It's a pain to understand the code and what it is doing. If possible, increase the number of comments to increase the understanding, even for developers with less background in RGW.

File: src/rgw/rgw_sal_simplefile.cc [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+  ldout(ctx(), 10) << __func__ << ": TODO" << dendl;
+  return "";
+}
+
+std::unique_ptr<LuaScriptManager> SimpleFileStore::get_lua_script_manager() {
+  ldout(ctx(), 10) << __func__ << ": TODO" << dendl;
+  return std::make_unique<UnsupportedLuaScriptManager>();
+}
+
+std::unique_ptr<RGWRole> SimpleFileStore::get_role(
```
Comment (votdev) [inline src/rgw/rgw_sal_simplefile.cc:259]: This is related to https://github.com/aquarist-labs/s3gw-core/issues/61.

File: src/rgw/rgw_sal_simplefile.cc [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    std::string name, std::string tenant, std::string path,
+    std::string trust_policy, std::string max_session_duration_str,
+    std::multimap<std::string, std::string> tags) {
+  ldout(ctx(), 10) << __func__ << ": TODO" << dendl;
+  RGWRole *p = nullptr;
+  return std::unique_ptr<RGWRole>(p);
+}
+
+std::unique_ptr<RGWRole> SimpleFileStore::get_role(std::string id) {
```
Comment (votdev) [inline src/rgw/rgw_sal_simplefile.cc:268]: This is related to https://github.com/aquarist-labs/s3gw-core/issues/61.

File: src/rgw/rgw_sal_simplefile.cc [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  ldout(ctx(), 10) << __func__ << ": TODO" << dendl;
+  RGWRole *p = nullptr;
+  return std::unique_ptr<RGWRole>(p);
+}
+
+int SimpleFileStore::get_roles(const DoutPrefixProvider *dpp, optional_yield y,
+                               const std::string &path_prefix,
+                               const std::string &tenant,
+                               vector<std::unique_ptr<RGWRole>> &roles) {
+  ldpp_dout(dpp, 10) << __func__ << ": TODO" << dendl;
```
Comment (votdev) [inline src/rgw/rgw_sal_simplefile.cc:278]: This is related to https://github.com/aquarist-labs/s3gw-core/issues/61.

File: src/rgw/store/simplefile/bucket.cc [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+  SimpleFileBucket::Meta meta;
+  JSONDecoder::decode_json("meta", meta, &bucket_meta_parser); 
+  lsfs_dout(dpp, 10) "bucket name: " << meta.info.bucket.get_key() << dendl;
+
+  info = meta.info;
+  // multipart = meta.multipart;
+
+  load_object_map(dpp);
+
+  auto f = new JSONFormatter(true);
```
Comment (votdev) [inline src/rgw/store/simplefile/bucket.cc:214]: Does it make sense to use this code? This way you do not need to take care to delete the new instantiated object.
```suggestion
  JSONFormatter f(true);
```

Comment (votdev) [inline src/rgw/store/simplefile/bucket.cc:214]: `delete f;` is missing after usage.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/rgw/store/simplefile/writer.cc [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+int SimpleFileAtomicWriter::prepare(optional_yield y) {
+  lsfs_dout(dpp, 10) << ": unimplemented, return success." << dendl;
+  // TODO: create meta file for this new object
+  return 0;
+}
+
+int SimpleFileAtomicWriter::process(bufferlist &&data, uint64_t offset) {
+  lsfs_dout(dpp, 10) << "data len: " << data.length()
+                     << ", offset: " << offset << dendl;
+
```
Comment (0xavi0) [inline src/rgw/store/simplefile/writer.cc:55]: Thinking about performance and given this is called from a loop and the writer instance is the same....
Maybe we could reuse the ofstream instance and even avoid the seekp call, flush and close.
Could we just call flush and close in the complete method?

Also avoid recalculating the object_path, as it's going to be the same for every call in the loop.

File: src/rgw/store/simplefile/writer.cc [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  lsfs_dout(dpp, 10) << "write to object at " << object_path << dendl;
+
+  auto mode = std::ofstream::binary|std::ofstream::out|std::ofstream::app;
+  std::ofstream ofs(object_path, mode);
+  ofs.seekp(offset);
+  data.write_stream(ofs);
+  ofs.flush();
+  ofs.close();
+  bytes_written += data.length();
```
Comment (0xavi0) [inline src/rgw/store/simplefile/writer.cc:66]: ...

File: src/rgw/store/simplefile/writer.cc [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  std::filesystem::path object_path = obj.get_data_path();
+
+  lsfs_dout(dpp, 10) << "write to object at " << object_path << dendl;
+
+  auto mode = std::ofstream::binary|std::ofstream::out|std::ofstream::app;
+  std::ofstream ofs(object_path, mode);
+  ofs.seekp(offset);
+  data.write_stream(ofs);
+  ofs.flush();
+  ofs.close();
```
Comment (0xavi0) [inline src/rgw/store/simplefile/writer.cc:65]: Thinking about performance and given this is called from a loop and the writer instance is the same....
Maybe we could reuse the ofstream instance and even avoid the seekp call, flush and close.
Could we just call flush and close in the complete method?

Also avoid recalculating the object_path, as it's going to be the same for every call in the loop.

Comment (jecluis) [reply src/rgw/store/simplefile/writer.cc:68]: Oh yeah, there's a bunch of those little things one can do. Most of these result from a "path of least resistance" kind of approach unfortunately (i.e., laziness). :(

Comment (0xavi0) [reply src/rgw/store/simplefile/writer.cc:68]: Good thing is when we change all those things... maybe our performance is not that bad compared to MinIO ;)

File: src/rgw/store/simplefile/user.cc [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    bool exclusive,
+    bool obj_lock_enabled,
+    bool *existed,
+    req_info &req_info,
+    std::unique_ptr<Bucket> *bucket,
+    optional_yield y
+) {
+  ceph_assert(bucket != nullptr);
+
+  auto f = new JSONFormatter(true);
```
Comment (votdev) [inline src/rgw/store/simplefile/user.cc:152]: Use `JSONFormatter f(true)` like https://github.com/aquarist-labs/ceph/pull/8/files#diff-2b9f89f3573abb0614d72a4533ce2140b34fb4222046482dcab3d848761d8da8R289 or ensure to delete the object after use.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/rgw/store/simplefile/user.cc (hunk 11 above)
Comment (jecluis) [reply src/rgw/store/simplefile/user.cc:152]: this needs to go away anyway. I don't think we can allow ourselves to write stuff as json anway.

**Comment by jecluis** [comment]
Okay, even though there are things that need to be changed (e.g., what @votdev mentioned, which needs to become encoded bufferlists instead of JSON), I'm now reasonably happy with this patch set.

I'd recommend merging, so we can continue development and possibly dividing tasks instead of keeping this big PR that is not going to stop growing.

**Comment by 0xavi0** [comment]
I agree, we can merge this if it's not breaking anything for `DBStore` so we can keep improving it. 
Anyway, if you don't set the store to be simplefile it should not interfere with the rest of stores.

**Review by 0xavi0 — APPROVED** [review]
I've tested the smoke test against `DBStore` and it passed. 
LGTM

**Comment by mattbenjamin** [comment]
hi @jecluis ;  this is interesting work;  have you considered presenting on it to the upstream RGW team (e.g., at our weekly refactoring meeting)?

**Comment by jecluis** [comment]
Hey @mattbenjamin -- we have definitely been considering when the best time to reach out to the rgw team would be. We have been fiddling with different ideas for the on-disk format though (e.g., see #17 and #18), and I think we've been postponing it until we feel confident enough on something worth have a discussion over. This PR we're chatting on had terrible performance (slightly better than dbstore at the time, but hardly so).

Would the refactoring meeting be the one marked as the RGW bug scrub (Thurs 14h UTC) on the community calendar? I'd be happy to discuss it soon.

**Comment by mattbenjamin** [comment]
hi @jecluis ;  no, the refactoring meeting is every wednesday, at 11:30 am, us eastern--we've been elevating this as "the" upstream steering meeting for a couple years, and we have other zipper folks present regularly (e.g., seagate/motr, intel daos)--it uses the regular meeting link;  you are invited :)

**Comment by mattbenjamin** [comment]
one of our (well, my) ideas was to use (in progress) zipper filter drivers to layer a high-performance i/o path for object/file data, on top of the existing dbstore (which could itself be speeded up by retargeting to an alternate backend);  I'd be interested in your feedback on same

**Comment by jecluis** [comment]
> hi @jecluis ; no, the refactoring meeting is every wednesday, at 11:30 am, us eastern--we've been elevating this as "the" upstream steering meeting for a couple years, and we have other zipper folks present regularly (e.g., seagate/motr, intel daos)--it uses the regular meeting link; you are invited :)

Thanks :) I'll try attending this week.

> one of our (well, my) ideas was to use (in progress) zipper filter drivers to layer a high-performance i/o path for object/file data, on top of the existing dbstore (which could itself be speeded up by retargeting to an alternate backend); I'd be interested in your feedback on same

@mattbenjamin Well, our experience with dbstore was great from a PoC point of view, but once we started looking into its performance it started lacking. And while I'm sure a lot of those issues could be solvable, one thing that became apparent was that it would always trail behind a data path targeting a filesystem for larger objects. I don't have a strong opinion there, but I think the only way to pull off relying on sqlite for object storage is by chunking objects into smaller blobs. But I'm more curious to understand why relying on sqlite for data? One of our current approaches is using it nonetheless, but solely for metadata, and leave the data path hitting the filesystem directly.

**Comment by mattbenjamin** [comment]
The value of sqlite was in a) compactness and b) transactional semantics.  Moreover, the dbstore implementation has been designed so that it could be retargeted to (an?) other sql-compliant backend--which could be a more scalable single node system, clustered system, etc.  Your suggestion to use dbstore just for metadata and file i/o for an object data path is exactly what I was suggesting above ("use (in progress) zipper filter drivers to layer a high-performance i/o path for object/file data, on top of the existing dbstore").  I think this is a promising direction.

**Pull request merged**